File: dart-wrapper/cardinal_sdk/ios/Classes/InitializersPlugin.swift
import Flutter
import CardinalDartSdkSupportLib

class InitializersPlugin {
	public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
		let args = call.arguments as! [String: String]
		let methodName = call.method
		
		let completed = InitializersPlugin.dispatch(methodName: methodName, parameters: args) { success, errorCode, errorMessage, errorDetails in
			if (errorCode != nil){
				result(FlutterError(code: errorCode!, message: errorMessage, details: errorDetails))
			}
			else {
				result(success)
			}
		}
		
		if !completed {
			result(FlutterMethodNotImplemented)
		}
	}
	
	private static func dispatch(
		methodName: String,
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Bool {
		switch methodName {
		case "initialize": initialize(parameters: parameters, resultCallback: resultCallback)
		case "initializeBase": initializeBase(parameters: parameters, resultCallback: resultCallback)
		case "switchGroup": switchGroup(parameters: parameters, resultCallback: resultCallback)
		case "baseSwitchGroup": baseSwitchGroup(parameters: parameters, resultCallback: resultCallback)
		case "initializeWithAuthProcess": initializeWithAuthProcess(parameters: parameters, resultCallback: resultCallback)
		case "completeAuthentication": completeAuthentication(parameters: parameters, resultCallback: resultCallback)
		default: return false
		}
		return true
	}
	
	private static func initialize(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.initializeSdk(
			dartResultCallback: resultCallback,
			applicationIdString: parameters["applicationId"]!,
			baseUrlString: parameters["baseUrl"]!,
			authenticationMethodString: parameters["authenticationMethod"]!,
			storageFacade: StorageHelper.shared.getStorage(options: parameters["storageOptions"]!),
			optionsString: parameters["options"]!
		)
	}

	private static func initializeBase(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.initializeBaseSdk(
			dartResultCallback: resultCallback,
			applicationIdString: parameters["applicationId"]!,
			baseUrlString: parameters["baseUrl"]!,
			authenticationMethodString: parameters["authenticationMethod"]!,
			optionsString: parameters["options"]!
		)
	}

	private static func switchGroup(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.switchGroup(
			dartResultCallback: resultCallback,
			sdkId: parameters["sdkId"]!,
			groupId: parameters["groupId"]!
		)
	}

	private static func baseSwitchGroup(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.baseSwitchGroup(
			dartResultCallback: resultCallback,
			sdkId: parameters["sdkId"]!,
			groupId: parameters["groupId"]!
		)
	}

	private static func initializeWithAuthProcess(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.initializeWithAuthProcess(
			dartResultCallback: resultCallback,
			applicationIdString: parameters["applicationId"]!,
			baseUrlString: parameters["baseUrl"]!,
			messageGatewayUrlString: parameters["messageGatewayUrl"]!,
			externalServicesSpecIdString: parameters["externalServicesSpecId"]!,
			processIdString: parameters["processId"]!,
			userTelecomTypeString: parameters["userTelecomType"]!,
			userTelecomString: parameters["userTelecom"]!,
			captchaOptionsString: parameters["captchaOptions"]!,
			baseStorage: StorageHelper.shared.getStorage(options: parameters["storageOptions"]!),
			authenticationProcessTemplateParametersString: parameters["authenticationProcessTemplateParameters"]!,
			optionsString: parameters["options"]!
		)
	}
	
	private static func completeAuthentication(
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?,
			String?
		) -> Void
	) -> Void {
		Initializers.shared.completeAuthentication(
			dartResultCallback: resultCallback,
			authenticationStepId: parameters["authenticationStepId"]!,
			validationCodeString: parameters["validationCode"]!
		)
	}
}

File: dart-wrapper/cardinal_sdk/android/src/main/kotlin/com/icure/cardinal_sdk/InitializersPlugin.kt
package com.icure.cardinal_sdk

import android.content.Context
import com.icure.cardinal.sdk.dart.Initializers
import com.icure.cardinal.sdk.dart.options.StorageHelper
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class InitializersPlugin(
    private val applicationContext: Context
) : MethodChannel.MethodCallHandler {
    @Suppress("UNCHECKED_CAST")
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (
            !dispatch(
                methodName = call.method,
                parameters = (call.arguments as Map<String, String>?).orEmpty()
            ) { success, errorCode, errorMessage, errorDetail ->
                if (errorCode != null)
                    result.error(errorCode, errorMessage, errorDetail)
                else
                    result.success(success)
            }
        ) result.notImplemented()
    }

    private fun dispatch(
        methodName: String,
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ): Boolean = when (methodName) {
        "initialize" -> initialize(parameters, resultCallback)
        "initializeBase" -> initializeBase(parameters, resultCallback)
        "switchGroup" -> switchGroup(parameters, resultCallback)
        "baseSwitchGroup" -> baseSwitchGroup(parameters, resultCallback)
        "initializeWithAuthProcess" -> initializeWithAuthProcess(parameters, resultCallback)
        "completeAuthentication" -> completeAuthentication(parameters, resultCallback)
        else -> null
    }?.let { true } ?: false

    private fun initialize(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.initializeSdk(
            resultCallback,
            applicationIdString = parameters.getValue("applicationId"),
            baseUrlString = parameters.getValue("baseUrl"),
            authenticationMethodString = parameters.getValue("authenticationMethod"),
            storageFacade = StorageHelper.getStorage(
                parameters.getValue("storageOptions"),
                applicationContext
            ),
            optionsString = parameters.getValue("options")
        )
    }

    private fun initializeBase(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.initializeBaseSdk(
            resultCallback,
            applicationIdString = parameters.getValue("applicationId"),
            baseUrlString = parameters.getValue("baseUrl"),
            authenticationMethodString = parameters.getValue("authenticationMethod"),
            optionsString = parameters.getValue("options")
        )
    }

    private fun switchGroup(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.switchGroup(
            resultCallback,
            sdkId = parameters.getValue("sdkId"),
            groupId = parameters.getValue("groupId"),
        )
    }

    private fun baseSwitchGroup(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.baseSwitchGroup(
            resultCallback,
            sdkId = parameters.getValue("sdkId"),
            groupId = parameters.getValue("groupId"),
        )
    }

    private fun initializeWithAuthProcess(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.initializeWithAuthProcess(
            resultCallback,
            applicationIdString = parameters.getValue("applicationId"),
            baseUrlString = parameters.getValue("baseUrl"),
            messageGatewayUrlString = parameters.getValue("messageGatewayUrl"),
            externalServicesSpecIdString = parameters.getValue("externalServicesSpecId"),
            processIdString = parameters.getValue("processId"),
            userTelecomTypeString = parameters.getValue("userTelecomType"),
            userTelecomString = parameters.getValue("userTelecom"),
            captchaOptionsString = parameters.getValue("captchaOptions"),
            baseStorage = StorageHelper.getStorage(
                parameters.getValue("storageOptions"),
                applicationContext
            ),
            authenticationProcessTemplateParametersString = parameters.getValue("authenticationProcessTemplateParameters"),
            optionsString = parameters.getValue("options"),
        )
    }

    private fun completeAuthentication(
        parameters: Map<String, String>,
        resultCallback: (
            String?,
            String?,
            String?,
            String?,
        ) -> Unit,
    ) {
        Initializers.completeAuthentication(
            resultCallback,
            authenticationStepId = parameters.getValue("authenticationStepId"),
            validationCodeString = parameters.getValue("validationCode"),
        )
    }
}